cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1939,7 +1940,8 @@ class S3FileSystem::Impl : public std::enable_shared_from_this<S3FileSystem::Imp
         return false;
       }
       RETURN_NOT_OK(self->CheckNestingDepth(nesting_depth));
-      return select.recursive && nesting_depth <= select.max_recursion;
+      return select.type == SelectorType::Directory &&
```

Comment:
Same here, I'm not sure how we reach this point unless we are already scanning a directory so this extra check seems redundant.